controller.go

```diff
@@ -469,3 +475,79 @@ func upgradePVFrom12(pv *PV) (deleted bool, err error) {
 	}
 	return false, nil
 }
+
+func provisionClaim(pvc *api.PersistentVolumeClaim) error {
+	// Provisioning can take some time. Start it in a separate goroutine.
+	// Keep a map of running provisioners and start new provisioning only if
+	// it's not already running. There will be some locks around, these are
+	// left out to implementation.
+	//
+
+	if !runningProvisioners.isRunning(pvc) {
+		runningProvisioners.markAsRunning(pvc)
+		go doProvisionClaim(pvc)
+	}
+	return nil
+}
+
+func doProvisionClaim(pvc *api.PersistentVolumeClaim) {
+	// On exit, mark the provisioner for this claim as finished. We don't want
+	// multiple doProvisionClaims running in parallel for the same claim.
+	defer runningProvisioners.markAsFinished(pvc)
+
+	// We're the only provisioner running for the PVC now. A previous
+	// doProvisionClaim may just have finished while we were waiting for the
+	// locks. Check that PV (with deterministic name) hasn't been provisioned
+	// yet.
+	pvName := calculateProvisionedPVNameForClaim(claim)
+	pv := kubeClient.Core().PersistentVolumes().Get(pvName)
+	if pv != nil {
+		// Previous doProvisionClaim has provisioned something, let the usual
+		// binding happen.
+		return
+	}
+
+	// Now we are sure that we need to provision something.
+	plugin := findProvisionerPluginForPV(claim)
+	if plugin = nil {
+		// Emit event that provisioning failed.
+		// Do not modify the claim and let the next syncClaim() try again.
+		return
+	}
+
+	// Provision the volume. Plugin is responsible to return valid PV object.
+	pv, err := plung.Provision(claim)
+	if err != nil {
+		// Emit event that provisioning failed.
+		// Do not modify the claim and let the next syncClaim() try again.
+		return
+	}
+
+	// NOTE: if the controller crashes here (or exits or gets SIGTERM), we will
+	// leave orphaned storage asset without having a PV for it in Kubernetes.
```

Comment:
Yeah, I don't think there's any way to prevent these kind of leaks as a guarantee, across all arbitrary cloud providers.  Seems like a fact of life that we may wind up leaking here.  We should call that out.